mars/services/web/core.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 1999-2020 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import cloudpickle
+import inspect
+import logging
+import sys
+import threading
+import time
+
+from collections import OrderedDict
+from tornado.httpclient import AsyncHTTPClient
+from .supervisor import MarsRequestHandler
+
+logger = logging.getLogger(__name__)
+
+
+def serialize(obj):
+    from mars.serialization import serialize
+    return cloudpickle.dumps(serialize(obj))
+
+
+def deserialize(binary):
+    from mars.serialization import deserialize
+    return deserialize(*(cloudpickle.loads(binary)))
+
+
+_web_address_local = threading.local()
+
+
+def set_web_address(web_address):
+    _web_address_local.web_address = web_address
+
+
+def get_web_address():
+    assert hasattr(_web_address_local, 'web_address'), 'Please set web_address first before get web address'
+    return _web_address_local.web_address
+
+
+class _HandlerException(Exception):
+
+    def __init__(self, exc_type, exc_value: BaseException, exc_traceback):
+        self.exc_type = exc_type
+        self.exc_value = exc_value
+        self.exc_traceback = exc_traceback
+
+
+_expired_sec = 60
+_keep_alive_interval = 3
+
+
+class ApiRegistry:
+
+    def __init__(self, expired_sec):
+        self._apis = dict()
```

Comment:
Every API instance belongs to a session, Is it possible to use session id as the key of api instance?
```python
{"session id": {TaskAPI: instance}}
```
Then, all web client shares the same instance for the same API class. We can recycle the session instances through last idle time.

Reply:
Removed api instance